### DISCUSSION TITLE
Debug vscode in attach mode

### DISCUSSION THREAD

**Discussion by taisso** [adonisjs/core#3896] [Help] [answered]
Hello, I wanted to know how I could debug in "attach" mode in adonisV5 using vscode, could someone help me? configuration I tried.

```
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Attach",
      "port": 3333,
      "restart": true,
      "request": "attach",
      "skipFiles": [
        "<node_internals>/**"
      ],
      "type": "node"
    },
  ]
}

```

**Comment by WeyIin** [comment, marked as answer]
Hi @taisso ,

I have configured one for you:

```
{
    "version": "0.2.0",
    "configurations": [
      {
        "type": "node",
        "request": "launch",
        "name": "Debug Adonis",
        "runtimeExecutable": "npm",
        "runtimeArgs": [
          "run",
          "devInspect"
        ],
        "internalConsoleOptions": "openOnSessionStart"
      },
      {
        "type": "node",
        "request": "attach",
        "name": "attach",
        "port": 9229
      }
    ]
  }
```

package.json:
```
  "scripts": {
    "devInspect": "node ace serve --watch --node-args=\"--inspect\"",
  }
```

Right, so while the first one "debug adonis" starts the server and attaches the debugger, the second one you use when your instance is already running. 

Hope that helps.

**Reply by taisso** [reply]
it worked, thanks for the help!

**Comment by davidharting** [comment]
The accepted answer helped me out, thanks so much @WeyIin!
I used your format to to also set up a profile to debug tests so I figured I'd share it out here. 

The test debug profile here will run all tests, breaking on whatever breakpoints you add. This could be improved to easily debug an individual test file. 

<details>

<summary>package.json</summary>
Just including the relevant excerpt

```json
  "scripts": {
    "dev:debug": "node ace serve --watch --node-args=\"--inspect\"",
    "test:debug": "node ace test --node-args=\"--inspect\"",
  },
```

</details>

<details>
<summary>.vscode/launch.json</summary>
Including it in it's entirety

```json
{
  "configurations": [
    {
      "type": "node",
      "request": "launch",
      "name": "Debug Web Server",
      "runtimeExecutable": "npm",
      "runtimeArgs": ["run", "dev:debug"],
      "internalConsoleOptions": "openOnSessionStart"
    },
    {
      "type": "node",
      "request": "launch",
      "name": "Debug Tests",
      "runtimeExecutable": "npm",
      "runtimeArgs": ["run", "test:debug"],
      "internalConsoleOptions": "openOnSessionStart"
    },
    {
      "type": "node",
      "request": "attach",
      "name": "Attach to Debugger",
      "port": 9229
    }
  ]
}

```

</details>

**Reply by davidharting** [reply]
I was able to get **Debug Current Test File** set up as well.

1. npm script

`"test:debug:file": "node ace test --node-args=\"--inspect\" --files",`

2. Configuration from .vscode/launch.json
```json
{
  "type": "node",
  "request": "launch",
  "name": "Debug Current Test File",
  "runtimeExecutable": "npm",
  "runtimeArgs": ["run", "test:debug:file", "${relativeFile}"],
  "internalConsoleOptions": "openOnSessionStart"
},
```

**Comment by smilingkylan** [comment]
@davidharting You're a live-saver... that worked first time for me when I've never really been able to get it to work well with Adonis v5. Muito obrigado!
